{
 "cells": [
  {
   "cell_type": "raw",
   "id": "3184d6d7-c1b7-4eda-962a-1210638eef87",
   "metadata": {},
   "source": [
    "# Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "77d844b2-a0d5-47ae-8c1f-e74fc3705370",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "raw",
   "id": "709a4b07-d53e-4d5a-bea0-52ecbe399043",
   "metadata": {},
   "source": [
    "# Set folder location and default info\n",
    "## The file *.txt contain the name of the isolates to be included in the final report\n",
    "## Raw folder contains the folder where the files obtained from MGAP_viz are saved\n",
    "## Batch contain corresponds to the name of the batch to be included in the final table, this name is useful in case the same sample is included in several batches. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "57e0ff85-3d67-4ae7-a56f-cde58e9ba2eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "275"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('/home/jugalde/germlab/genome_test/samples.txt') as f:\n",
    "    sample_ID = f.read().splitlines()\n",
    "\n",
    "raw_folder = '/home/jugalde/germlab/genome_test/consolidated//'\n",
    "batch = 'mrsa_genomes'\n",
    "\n",
    "len(sample_ID)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c4c2d0a3-d550-4a35-9e78-f668648c108e",
   "metadata": {},
   "source": [
    "# The following script summarizes the information using several files obtained from the MGAP_viz folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5b370eee",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Info about the number of reads \n",
    "\n",
    "fastp = pd.read_table(raw_folder+'fastp_results.tsv',delimiter = '\\t')\n",
    "fastp_clean = pd.DataFrame().assign(sample_ID = fastp[\"sample\"], n_reads_before_trimming = fastp[\"reads_before\"],\n",
    "                                   n_read_after_trimming = fastp[\"reads_after\"])\n",
    "\n",
    "## Info about the data obtained from BRAKEN\n",
    "\n",
    "# braken = pd.read_table(raw_folder+'bracken_results.tsv',delimiter = '\\t')\n",
    "# braken_clean = pd.DataFrame(columns=['sample_ID', 'braken_species_first', 'braken_taxonomy_id_first',\n",
    "#                                      'braken_fraction_reads_first', 'braken_species_second',\n",
    "#                                      'braken_taxonomy_id_second', 'braken_fraction_reads_second'])\n",
    "\n",
    "# ## Selects the two highest hit from bracken and organizes the info\n",
    "# for i in range(len(sample_ID)):\n",
    "#     braken_temp = braken[braken[\"sample\"] == sample_ID[i]].sort_values(by='fraction_total_reads', ascending = False)\n",
    "    \n",
    "#     if braken_temp.shape[0] > 0:\n",
    "#         braken_species_first = braken_temp.iloc[0]['name']\n",
    "#         braken_taxonomy_id_first = braken_temp.iloc[0]['taxonomy_id']\n",
    "#         braken_fraction_reads_first = braken_temp.iloc[0]['fraction_total_reads']\n",
    "#     else:\n",
    "#         braken_species_first = None\n",
    "#         braken_taxonomy_id_first = None\n",
    "#         braken_fraction_reads_first = None\n",
    "\n",
    "#     if braken_temp.shape[0] > 1:\n",
    "#         braken_species_second = braken_temp.iloc[1]['name']\n",
    "#         braken_taxonomy_id_second = braken_temp.iloc[1]['taxonomy_id']\n",
    "#         braken_fraction_reads_second = braken_temp.iloc[1]['fraction_total_reads']\n",
    "#     else:\n",
    "#         braken_species_second = None\n",
    "#         braken_taxonomy_id_second = None\n",
    "#         braken_fraction_reads_second = None\n",
    "     \n",
    "#     braken_temp_row = [sample_ID[i],braken_species_first,braken_taxonomy_id_first,braken_fraction_reads_first,\n",
    "#                   braken_species_second,braken_taxonomy_id_second,braken_fraction_reads_second]\n",
    "\n",
    "#     braken_clean.loc[len(braken_clean)] = braken_temp_row\n",
    "\n",
    "## Info about the assembly quality\n",
    "\n",
    "qc_assembly = pd.read_table(raw_folder+'qc_assembly.tsv',delimiter = '\\t')\n",
    "qc_assembly_clean = pd.DataFrame().assign(sample_ID = qc_assembly[\"sample\"],\n",
    "#                                          contigs_0 = qc_assembly[\"contigs_0\"],\n",
    "#                                          contigs_10k = qc_assembly[\"contigs_10k\"],\n",
    "#                                          contigs_50k = qc_assembly[\"contigs_50k\"],\n",
    "                                          length_total_0 = qc_assembly[\"contigs_total\"],\n",
    "                                          contigs = qc_assembly[\"length_50k\"],\n",
    "                                          length_total = qc_assembly[\"largest_contig\"],\n",
    "                                          largest_contig = qc_assembly[\"length_total\"],\n",
    "                                          GC_content = qc_assembly[\"gc\"],\n",
    "                                          N50 = qc_assembly[\"n50\"])\n",
    "\n",
    "## Info about the MLST results\n",
    "\n",
    "mlst = pd.read_table(raw_folder+'mlst_results.tsv',delimiter = '\\t')\n",
    "mlst_clean = mlst.rename(columns = {'sample':'sample_ID','species':\"MLST_species\"})\n",
    "\n",
    "## Info about the AMR finder results\n",
    "\n",
    "amrfinder = pd.read_table(raw_folder+'amrfinder_results.tsv',delimiter = '\\t')\n",
    "\n",
    "amrfinder_carb_clean = pd.DataFrame(columns=['sample_ID', 'carbapenemases', 'IMP', 'KPC', 'NDM', 'VIM',\n",
    "                                             'others_carbapenem', 'point_mut_carbapenem'])\n",
    "\n",
    "## Organizes the AMR info starting by Carbapenemases\n",
    "\n",
    "for i in range(len(sample_ID)):\n",
    "    amrfinder_temp = amrfinder[amrfinder[\"sample\"] == sample_ID[i]]\n",
    "\n",
    "    amrfinder_temp_carb = amrfinder_temp[amrfinder_temp[\"Subclass\"] == \"CARBAPENEM\"]\n",
    "    amrfinder_temp_carb_amr = amrfinder_temp_carb[amrfinder_temp_carb[\"Element subtype\"] == \"AMR\"]\n",
    "    amrfinder_temp_carb_point = amrfinder_temp_carb[amrfinder_temp_carb[\"Element subtype\"] == \"POINT\"]\n",
    "\n",
    "    amrfinder_temp_carb_IMP_table = amrfinder_temp_carb_amr[amrfinder_temp_carb_amr['Gene symbol'].str.match(\"blaIMP\")]\n",
    "    amrfinder_temp_carb_KPC_table = amrfinder_temp_carb_amr[amrfinder_temp_carb_amr['Gene symbol'].str.match(\"blaKPC\")]\n",
    "    amrfinder_temp_carb_NDM_table = amrfinder_temp_carb_amr[amrfinder_temp_carb_amr['Gene symbol'].str.match(\"blaNDM\")]\n",
    "    amrfinder_temp_carb_VIM_table = amrfinder_temp_carb_amr[amrfinder_temp_carb_amr['Gene symbol'].str.match(\"blaVIM\")]\n",
    "    amrfinder_temp_carb_others_table = amrfinder_temp_carb_amr[~amrfinder_temp_carb_amr['Gene symbol'].\n",
    "                                                               str.match(\"blaKPC|blaNDM|blaIMP|blaVIM\")]\n",
    "    if amrfinder_temp_carb_amr.shape[0] > 0:\n",
    "        carbapenemases_temp = \"Yes\"\n",
    "    else:\n",
    "        carbapenemases_temp = \"No\"\n",
    "\n",
    "    if amrfinder_temp_carb_IMP_table.shape[0] > 0:\n",
    "        IMP_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_carb_IMP_table[\"Name of closest sequence\"].to_list()]\n",
    "        IMP_temp = \", \".join(IMP_temp)\n",
    "    else:\n",
    "        IMP_temp = None\n",
    "\n",
    "    if amrfinder_temp_carb_KPC_table.shape[0] > 0:\n",
    "        KPC_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_carb_KPC_table[\"Name of closest sequence\"].to_list()]\n",
    "        KPC_temp = \", \".join(KPC_temp)\n",
    "    else:\n",
    "        KPC_temp = None\n",
    "        \n",
    "    if amrfinder_temp_carb_NDM_table.shape[0] > 0:\n",
    "        NDM_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_carb_NDM_table[\"Name of closest sequence\"].to_list()]\n",
    "        NDM_temp = \", \".join(NDM_temp)\n",
    "    else:\n",
    "        NDM_temp = None\n",
    "\n",
    "    if amrfinder_temp_carb_VIM_table.shape[0] > 0:\n",
    "        VIM_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_carb_VIM_table[\"Name of closest sequence\"].to_list()]\n",
    "        VIM_temp = \", \".join(VIM_temp)\n",
    "    else:\n",
    "        VIM_temp = None\n",
    "\n",
    "    if amrfinder_temp_carb_others_table.shape[0] > 0:\n",
    "        others_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_carb_others_table[\"Name of closest sequence\"].to_list()]\n",
    "        others_temp = \", \".join(others_temp)\n",
    "    else:\n",
    "        others_temp = None\n",
    "\n",
    "    if amrfinder_temp_carb_point.shape[0] > 0:\n",
    "        point_temp = \", \".join(amrfinder_temp_carb_point[\"Gene symbol\"].to_list())\n",
    "    else:\n",
    "        point_temp = None\n",
    "\n",
    "    amrfinder_temp_row = [sample_ID[i], carbapenemases_temp, IMP_temp, KPC_temp, NDM_temp, VIM_temp, others_temp,\n",
    "                          point_temp]\n",
    "\n",
    "    amrfinder_carb_clean.loc[len(amrfinder_carb_clean)] = amrfinder_temp_row\n",
    "\n",
    "## Organizes the AMR info for ESBLs and ESBLs of interest\n",
    "\n",
    "amrfinder_esbl_clean = pd.DataFrame(columns=['sample_ID', 'ESBL', 'CTX-M', 'SHV', 'ESBL_TEM', 'others_ESBL',\n",
    "                                       'point_mut_beta_lactam', 'all_TEM'])\n",
    "\n",
    "for i in range(len(sample_ID)):\n",
    "    amrfinder_temp = amrfinder[amrfinder[\"sample\"] == sample_ID[i]]\n",
    "    amrfinder_temp_esbl = amrfinder_temp[amrfinder_temp[\"Name of closest sequence\"].\n",
    "                                         str.contains(\"extended-spectrum|broad-spectrum\")]\n",
    "    amrfinder_temp_esbl = amrfinder_temp_esbl[amrfinder_temp_esbl[\"Name of closest sequence\"].\n",
    "                                              str.contains(\"beta-lactamase\")]\n",
    "    amrfinder_temp_esbl_amr = amrfinder_temp_esbl[amrfinder_temp_esbl[\"Element subtype\"] == \"AMR\"]\n",
    "    amrfinder_temp_esbl_point = amrfinder_temp_esbl[amrfinder_temp_esbl[\"Element subtype\"] == \"POINT\"]\n",
    "\n",
    "    amrfinder_temp_esbl_CTX_M_table = amrfinder_temp_esbl_amr[amrfinder_temp_esbl_amr['Gene symbol'].str.match(\"blaCTX-M\")]\n",
    "    amrfinder_temp_esbl_SHV_table = amrfinder_temp_esbl_amr[amrfinder_temp_esbl_amr['Gene symbol'].str.match(\"blaSHV\")]\n",
    "    amrfinder_temp_esbl_TEM_table = amrfinder_temp_esbl_amr[amrfinder_temp_esbl_amr['Gene symbol'].str.match(\"blaTEM\")]\n",
    "    amrfinder_temp_esbl_others_table = amrfinder_temp_esbl_amr[~amrfinder_temp_esbl_amr['Gene symbol'].\n",
    "                                                               str.match(\"blaCTX-M|blaSHV|blaTEM\")]\n",
    "    \n",
    "    amrfinder_temp_all_TEM_table = amrfinder_temp[amrfinder_temp['Gene symbol'].str.match(\"blaTEM\")]\n",
    "\n",
    "    if amrfinder_temp_esbl_amr.shape[0] > 0:\n",
    "        ESBL_temp = \"Yes\"\n",
    "    else:\n",
    "        ESBL_temp = \"No\"\n",
    "\n",
    "    if amrfinder_temp_esbl_CTX_M_table.shape[0] > 0:\n",
    "        CTX_M_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_esbl_CTX_M_table[\"Name of closest sequence\"].to_list()]\n",
    "        CTX_M_temp = \", \".join(CTX_M_temp)\n",
    "    else:\n",
    "        CTX_M_temp = None\n",
    "        \n",
    "    if amrfinder_temp_esbl_SHV_table.shape[0] > 0:\n",
    "        SHV_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_esbl_SHV_table[\"Name of closest sequence\"].to_list()]\n",
    "        SHV_temp = \", \".join(SHV_temp)\n",
    "    else:\n",
    "        SHV_temp = None\n",
    "        \n",
    "    if amrfinder_temp_esbl_TEM_table.shape[0] > 0:\n",
    "        ESBL_TEM_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_esbl_TEM_table[\"Name of closest sequence\"].to_list()]\n",
    "        ESBL_TEM_temp = \", \".join(ESBL_TEM_temp)\n",
    "    else:\n",
    "        ESBL_TEM_temp = None\n",
    "        \n",
    "    if amrfinder_temp_esbl_others_table.shape[0] > 0:\n",
    "        others_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_esbl_others_table[\"Name of closest sequence\"].to_list()]\n",
    "        others_temp = \", \".join(others_temp)\n",
    "    else:\n",
    "        others_temp = None\n",
    "    \n",
    "    if amrfinder_temp_esbl_point.shape[0] > 0:\n",
    "        point_temp = \", \".join(amrfinder_temp_esbl_point[\"Gene symbol\"].to_list())\n",
    "    else:\n",
    "        point_temp = None\n",
    "    \n",
    "    if amrfinder_temp_all_TEM_table.shape[0] > 0:\n",
    "        all_TEM_temp = [i.rsplit(\" \",1)[-1] for i in amrfinder_temp_all_TEM_table[\"Name of closest sequence\"].to_list()]\n",
    "        all_TEM_temp = \", \".join(all_TEM_temp)\n",
    "    else:\n",
    "        all_TEM_temp = None\n",
    "\n",
    "    amrfinder_temp_row = [sample_ID[i], ESBL_temp, CTX_M_temp, SHV_temp, ESBL_TEM_temp, others_temp, point_temp, all_TEM_temp]\n",
    "\n",
    "    amrfinder_esbl_clean.loc[len(amrfinder_esbl_clean)] = amrfinder_temp_row\n",
    "\n",
    "## Organizes the AMR info for mec and van\n",
    "\n",
    "amrfinder_mec_van_clean = pd.DataFrame(columns=['sample_ID', 'mecA', 'others_mec', 'VanA', 'VanB', 'others_van'])\n",
    "\n",
    "for i in range(len(sample_ID)):\n",
    "    amrfinder_temp = amrfinder[amrfinder[\"sample\"] == sample_ID[i]]\n",
    "\n",
    "    amrfinder_temp_mec_table = amrfinder_temp[amrfinder_temp['Gene symbol'].str.match(\"mec\")]\n",
    "    amrfinder_temp_meca_table = amrfinder_temp_mec_table[amrfinder_temp_mec_table['Gene symbol'].str.match(\"mecA\")]\n",
    "    amrfinder_temp_mec_others_table = amrfinder_temp_mec_table[~amrfinder_temp_mec_table['Gene symbol'].\n",
    "                                                               str.match(\"mecA\")]\n",
    "    \n",
    "    amrfinder_temp_van_table = amrfinder_temp[amrfinder_temp['Gene symbol'].str.match(\"van\")]\n",
    "    amrfinder_temp_vana_table = amrfinder_temp_van_table[amrfinder_temp_van_table['Gene symbol'].str.match(\"^vanA\")]\n",
    "    amrfinder_temp_vanb_table = amrfinder_temp_van_table[amrfinder_temp_van_table['Gene symbol'].str.match(\"^vanB\")]\n",
    "    amrfinder_temp_van_others_table = amrfinder_temp_van_table[~amrfinder_temp_van_table['Gene symbol'].\n",
    "                                                               str.match(\"vanA|vanB\")]\n",
    "    if amrfinder_temp_meca_table.shape[0] > 0:\n",
    "        meca_temp = \", \".join(amrfinder_temp_meca_table[\"Gene symbol\"].to_list())\n",
    "    else:\n",
    "        meca_temp = None\n",
    "\n",
    "    if amrfinder_temp_mec_others_table.shape[0] > 0:\n",
    "        mec_others_temp = \", \".join(amrfinder_temp_mec_others_table[\"Gene symbol\"].to_list())\n",
    "    else:\n",
    "        mec_others_temp = None\n",
    "\n",
    "    if amrfinder_temp_vana_table.shape[0] > 0:\n",
    "        vana_temp = \", \".join(amrfinder_temp_vana_table[\"Gene symbol\"].to_list())\n",
    "    else:\n",
    "        vana_temp = None\n",
    "        \n",
    "    if amrfinder_temp_vanb_table.shape[0] > 0:\n",
    "        vanb_temp = \", \".join(amrfinder_temp_vanb_table[\"Gene symbol\"].to_list())\n",
    "    else:\n",
    "        vanb_temp = None\n",
    "\n",
    "    if amrfinder_temp_van_others_table.shape[0] > 0:\n",
    "        van_others_temp = \", \".join(amrfinder_temp_van_others_table[\"Gene symbol\"].to_list())\n",
    "    else:\n",
    "        van_others_temp = None\n",
    "        \n",
    "    amrfinder_temp_row = [sample_ID[i], meca_temp, mec_others_temp, vana_temp, vanb_temp, van_others_temp]\n",
    "\n",
    "    amrfinder_mec_van_clean.loc[len(amrfinder_mec_van_clean)] = amrfinder_temp_row\n",
    "    \n",
    "## Info about the SCCmec\n",
    "\n",
    "if os.path.isfile(raw_folder+'sccmec.tsv'):\n",
    "    sccmec = pd.read_table(raw_folder+'sccmec.tsv',delimiter = '\\t')    \n",
    "else:\n",
    "    sccmec = pd.DataFrame(columns=['sample', 'sccmec'])\n",
    "    \n",
    "sccmec_clean = pd.DataFrame(columns=['sample_ID', 'sccmec'])\n",
    "\n",
    "for i in range(len(sample_ID)):\n",
    "    sccmec_temp = sccmec[sccmec[\"sample\"] == sample_ID[i]]\n",
    "    if sccmec_temp.shape[0] == 1:\n",
    "        sccmec_data = sccmec_temp[\"sccmec\"].iloc[0]\n",
    "    else:\n",
    "        sccmec_data = \"NA\"\n",
    "    \n",
    "    sccmec_temp_row = [sample_ID[i], sccmec_data]\n",
    "    \n",
    "    sccmec_clean.loc[len(sccmec_clean)] =  sccmec_temp_row"
   ]
  },
  {
   "cell_type": "raw",
   "id": "fd9e7d9b-6abd-453a-9381-389f4d22b5ee",
   "metadata": {},
   "source": [
    "# Merges all tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "faf0d4b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "275"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_table = pd.DataFrame({'batch':([batch]*len(sample_ID)),'sample_ID':sample_ID})\n",
    "summary_table = pd.merge(summary_table,fastp_clean,how='left', on='sample_ID')\n",
    "#summary_table = pd.merge(summary_table,braken_clean,how='left', on='sample_ID')\n",
    "summary_table = pd.merge(summary_table,qc_assembly_clean,how='left', on='sample_ID')\n",
    "summary_table = pd.merge(summary_table,mlst_clean,how='left', on='sample_ID')\n",
    "summary_table = pd.merge(summary_table,amrfinder_carb_clean,how='left', on='sample_ID')\n",
    "summary_table = pd.merge(summary_table,amrfinder_esbl_clean,how='left', on='sample_ID')\n",
    "summary_table = pd.merge(summary_table,amrfinder_mec_van_clean,how='left', on='sample_ID')\n",
    "summary_table = pd.merge(summary_table,sccmec_clean,how='left', on='sample_ID')\n",
    "summary_table.shape[0]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d7721a55-5e84-48f8-84eb-0bd4ad48db94",
   "metadata": {},
   "source": [
    "# Exports table to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d738bae9-38ce-4731-873f-e9a06923895c",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_table.to_csv(\"/home/jugalde/germlab/genome_test/20250514_Summary.csv\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65ddfd9c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
